File: s-wm-shot.rb
#!/usr/bin/env ruby
# frozen_string_literal: true

require 'pathname'
require 'fileutils'
require 'optparse'

# TODO: add --verbose option

# Take a shot of the marvelous screen
class Screenshot
  def self.which?(executable)
    ENV['PATH'].split(File::PATH_SEPARATOR).any? do |directory|
      File.executable?(File.join(directory, executable.to_s))
    end
  end

  attr_reader :shotters

  def initialize(shotters = Shotters.new.all)
    @shotters = shotters
  end

  def full
    [shotter.exec, shotter.common, shotter.full, shotter.name].join ' '
  end

  def partial
    [shotter.exec, shotter.common, shotter.partial, shotter.name].join ' '
  end

  def info
    p <<~EOL # double quote document
      Shotters available: #{shotters.keys.join(', ')}
      location: #{shotter.name}
    EOL
  end

  private

  # list of all shotters found
  def shotters_available
    shotters.find_all do |shotter|
      x = shotter.last
      x.exec if Screenshot.which? x.exec
    end
  end

  # select default shotter
  def shotter
    shotters_available.first.last
  end
end

# All available shotters
class Shotters
  require 'ostruct'

  PREFERRED_FORMAT = 'png'
  CURRENT_TIME = Time.new.strftime '%d-%m-%y at %I-%M'

  def all
    {
      scrot: OpenStruct.new(exec: 'scrot',
                            full: '--focused --silent', partial: '--select --silent', common: '--border --quality=100',
                            name: screenshot_location),
      main: OpenStruct.new(exec: 'maim',
                           full: '', partial: '--select',
                           name: screenshot_location),
      flameshot: OpenStruct.new(exec: 'flameshot',
                                full: '', partial: '--select',
                                name: screenshot_location),
      grim: OpenStruct.new(exec: 'grim',
                           full: '', partial: '-g "$(slurp)"', common: '-t png -l 9',
                           name: screenshot_location)
    }
  end

  private

  def folder
    f = Pathname.new(Dir.home).join('Pictures', 'screenshots')
    FileUtils.mkdir_p f unless f.exist?

    f
  end

  def screenshot_location
    "'#{folder.join("Screenshot from #{CURRENT_TIME}.#{PREFERRED_FORMAT}")}'"
  end
end

options = OptionParser.new do |opts|
  shotter = Screenshot.new
  opts.banner = 'Usage: shot [options]'

  opts.on('-f', '--full', 'full screen shot') do
    system shotter.full
  end

  opts.on('-p', '--partial', 'partial screen shot') do
    system shotter.partial
  end

  opts.on('-i', '--info', 'shotters information') do
    puts shotter.info
  end
end

options.parse! ['--help'] if ARGV.empty?
options.parse!

File: s-wm-shot.go
/// 2>/dev/null ; gorun "$0" "$@" ; exit $?

package main

import (
	"flag"
	"fmt"
	"os"
	"os/exec"
	"path/filepath"
	"time"
)

type Shotter struct {
	exec    string
	full    string
	partial string
	common  string
	name    string
}

func (s *Shotter) full_end() []string {
	return []string{s.common, s.full, s.name}
}

func (s *Shotter) partial_end() []string {
	return []string{s.common, s.partial, s.name}
}

func main() {
	full, partial := shot_cli_parse()
	grim := grim()

	if *full == true {
		shot_command(grim.exec, grim.full_end())
	}

	if *partial == true {
		shot_command(grim.exec, grim.partial_end())
	}
}

func shot_location() string {
	home, err := os.UserHomeDir()

	if err != nil {
		fmt.Fprintln(os.Stderr, err.Error())
		os.Exit(1)
	}

	screen_folder := filepath.Join(home, "Pictures", "Screenshots")

	if _, err := os.Stat(screen_folder); err != nil {
		err = os.Mkdir(screen_folder, 0755)

		if err != nil {
			fmt.Fprintln(os.Stderr, err.Error())
			os.Exit(1)
		}
	}

	return screen_folder
}

func grim() Shotter {
	shot_name := filepath.Join(shot_location(), time.Now().Format(time.RFC3339)+".png")

	return Shotter{
		exec:    "grimshot",
		full:    "active",
		partial: "area",
		common:  "save",
		name:    shot_name,
	}
}

// command line arguments parser
func shot_cli_parse() (*bool, *bool) {
	full := flag.Bool("full", false, "full screen shot")
	partial := flag.Bool("partial", false, "partial screen shot")

	flag.Parse()

	flag.Usage = func() {
		fmt.Fprintf(flag.CommandLine.Output(), "Usage: volume [options]\n\n")
		flag.PrintDefaults()
	}

	if *full == false && *partial == false {
		flag.Usage()
		os.Exit(0)
	}

	return full, partial
}

func shot_command(cmd string, args []string) {
	command := exec.Command(cmd, args...)

	if err := command.Run(); err != nil {
		fmt.Fprintln(os.Stderr, err.Error())
		os.Exit(1)
	}

	if stdout, err := command.Output(); err != nil {
		fmt.Fprintln(os.Stderr, err.Error())
		os.Exit(1)
	} else {
		fmt.Println(string(stdout))
	}
}
